pkg/lib/cockpit-components-terminal.tsx

```diff
@@ -86,30 +85,137 @@
+ * fill its container.
  *
- * If the 'onTitleChanged' callback property is set, it will be called whenever
- * the title of the terminal changes.
+ * If the 'onTitleChanged' callback property is set, it will be called
+ * whenever the title of the terminal changes.
  *
- * Call focus() to set the input focus on the terminal.
+ * Call focus() on the Terminal component to set the input focus on
+ * the terminal, or reset() to clear it.
  *
  * Also it is possible to set up theme by property 'theme'.
  */
-export class Terminal extends React.Component {
-    constructor(props) {
+
+export class TerminalState {
+    terminal: Term;
+    element: HTMLDivElement;
+    channel: cockpit.Channel<string>;
+
+    constructor(channel: cockpit.Channel<string>) {
+        this.terminal = new Term({
+            cols: 80,
+            rows: 1,
+            cursorBlink: true,
+            fontSize: 16,
+            fontFamily: 'Menlo, Monaco, Consolas, monospace',
+            screenReaderMode: true,
+        });
+        this.terminal.loadAddon(new CanvasAddon());
+        this.element = document.createElement("div");
+        this.channel = channel;
+        this.#connectChannel(channel);
+    }
+
+    #connectChannel(channel: cockpit.Channel<string>) {
+        channel.addEventListener('message', (_event, data) => {
+            this.terminal.write(data);
+        });
+
+        this.terminal.onData(data => {
+            if (channel.valid) {
+                /* HACK: Ctrl+Space (and possibly other
+                 * characters) is a disaster: While it is U+00A0
+                 * in unicode, with an UTF-8 representation of
+                 * 0xC2A0, the "visible" string in JS is 0x00
+                 * (with TextEncoder, btoa(), and string
+                 * comparison). The internal representation
+                 * retains half of it, and trying to send it to
+                 * the websocket would result in a single 0xA0,
+                 * which is invalid UTF-8 (and causes the session
+                 * to crash). So intercept and ignore such broken
+                 * chars.  See
+                 * https://github.com/cockpit-project/cockpit/issues/21213 */
+                if (data === '\x00') {
+                    console.log("terminal: ignoring invalid input", data);
+                    return;
+                }
+                channel.send(data);
+            }
+        });
+
+        channel.addEventListener('close', (_event, options) => {
+            const term = this.terminal;
+            term.write('\x1b[31m' + (options.problem || 'disconnected') + '\x1b[m\r\n');
+            term.refresh(term.rows, term.rows);
+        });
+    }
+
+    resetChannel(channel: cockpit.Channel<string>) {
+        this.channel.close();
+        this.channel = channel;
+        this.#connectChannel(channel);
+    }
+
+    close() {
+        this.channel.close();
+        this.terminal.dispose();
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-21680-4a6f1830-20250815-070855-fedora-42-devel/Coverage/lcov/github-pr.diff.gcov.html)